### DISCUSSION TITLE
Why does Get() always return pointer type?

### DISCUSSION THREAD

**Discussion by setanarut** [yohamta/donburi#123] [Q&A] [answered]
Why does `Get()` always return pointer type? (`*float64`) Even though I created `float64` type component.

```Go
package main

import (
	"fmt"

	"github.com/yohamta/donburi"
)

func main() {
	w := donburi.NewWorld()
	health := donburi.NewComponentType[float64]()
	e := w.Entry(w.Create(health))
	health.SetValue(e, 5.0)
	fmt.Printf("%T\n", health.Get(e))
	// *float64
}

```

**Comment by yohamta0** [comment, marked as answer]
Hello, it's because donburi stores component data as pointer. I think we can add `GetValue()` method to get value instead of pointer. What do you think?

**Reply by setanarut** [reply]
I've read that using pointer type is slower for primitive types. (in terms of memory access speed). Also there is no donbori guide for singleton components. I don't know what's going on behind the scenes.

**Reply by yohamta0** [reply]
I see your point. Currently, donburi stores component data as `unsafe.pointer` here:

https://github.com/yohamta/donburi/blob/ba79e8311163d0e312c2d0cdd153ea07f33c819c/internal/storage/storage.go#L9-L16

I am a bit unsure about the impact of memory access speed, but perhaps we can experiment to see if it's significant enough to make changes. Regarding singleton components, Donburi does not have a dedicated API. Do you have any thoughts or ideas on that?

**Comment by setanarut** [comment]
> I see your point. Currently, donburi stores component data as `unsafe.pointer` here:
> 
> https://github.com/yohamta/donburi/blob/ba79e8311163d0e312c2d0cdd153ea07f33c819c/internal/storage/storage.go#L9-L16
> 
> I am a bit unsure about the impact of memory access speed, but perhaps we can experiment to see if it's significant enough to make changes. Regarding singleton components, Donburi does not have a dedicated API. Do you have any thoughts or ideas on that?

I changed my mind. I don't think it will affect speed for a few thousand Entities. In terms of syntax, it would be nice to have `GetValue()`.

---

> Donburi does not have a dedicated API. Do you have any thoughts or ideas on that?

A helper function can be added for the singleton component. For example, when `e1` is changed, all entities with the same component will also change. I think there is currently no way to do this without using `a` variable and `Set()` and `SetValue()` .

```Go
package main

import (
	"fmt"

	"github.com/yohamta/donburi"
)

func main() {
	a := "A"
	w := donburi.NewWorld()
	Str := donburi.NewComponentType[string]()

	e1 := w.Entry(w.Create(Str))
	e2 := w.Entry(w.Create(Str))
	Str.Set(e1, &a)
	Str.Set(e2, &a)

	Str.SetValue(e1, "B")

	fmt.Println(*Str.Get(e1))
	fmt.Println(*Str.Get(e2))
	// B
	// B
}


```

# Pseudo code in my mind


```Go
	Str := donburi.NewSingletonComponentType[string]("A")

	e1 := world.Entry(world.Create(Str))
	e2 := world.Entry(world.Create(Str))

	Str.SetValue(e1, "B")

	fmt.Println(*Str.Get(e1))
	fmt.Println(*Str.Get(e2))
	// B
	// B
```

**Reply by yohamta0** [reply]
Thanks for the example. The idea of a singleton component feature is indeed interesting. I believe it may be feasible to develop a `features/singleton` package which would enable arbitrary component to be treated as a singleton (transparently). What do you think?

**Reply by setanarut** [reply]
> Thanks for the example. The idea of a singleton component feature is indeed interesting. I believe it may be feasible to develop a `features/singleton` package which would enable arbitrary component to be treated as a singleton (transparently). What do you think?

I don't think this is a very important feature, but it is useful for legible code. Also, forcing to use global variables is against ECS design.

**Reply by yohamta0** [reply]
By the way, bevy, an game engine for Rust that uses ECS architecture, has a concept called [Resource](https://bevyengine.org/learn/quick-start/getting-started/resources/). This concept allows for handling data without an entity, and the data is automatically passed to systems.
